### ISSUE TITLE
Feature: Dask support in scipy.integrate functions 

### ISSUE THREAD

**Issue by guygma** [scipy/scipy#10087]
As far as I can tell, the ODE integrators available in ```scipy``` are not constructed with the option to use ```dask``` arrays. This imposes an unfortunate limitation when evolving a system of equations on a grid, despite that fact that ```dask``` appears an adequate solution. As a practical matter of implementation, how feasible would it be to make this addition to the code base? Perhaps there are a certain set of steps a core maintainer would recommend starting with and someone could take over along those lines if this feature is not given top priority.

**Comment by rgommers** [comment]
Dask is currently not a dependency of SciPy. It could become one in the future, however that requires discussion and a very good rationale. I'd like to know about how that's working out for other libraries (e.g. scikit-image), whether there are alternatives, what the benefit is, etc.

More importantly for now, I'm not sure I understand exactly what you mean by "evolving a system of equations on a grid". It sounds like more than a trivially parallelizable problem. A more detailed explanation would be a good start. How would you modify what functions exactly?

**Comment by guygma** [comment]
To clarify, I don't think having ```dask``` as a requirement would be necessary to implement this feature. It would simply be optional, because of how ```dask.array``` objects can be handled as a ```numpy.array``` for the most part. 

An example of the kind of application I have in mind is solving the diffusion equation, where the density, ```n(x; t)```, is propagated in time but also depends on the spatial gradient. Particularly when this is generalized to 3 dimensions, since the resolution of the gradient depends on the grid spacing it is easy to see how large arrays can become possible. The issue that I have personally run up against is a ```MemoryError``` induced by the ```solve_ivp``` algorithm while estimating the Jacobian matrix, which for a density array of size ```N``` requires an array of size ```N^2```. One potential solution might be to handle this error with a fallback to a ```dask.array``` if the package is available in the run-time environment. I think the most obvious change in that case is within ```_sparse_num_jac``` where the ```numpy.ndarray``` objects with shape ```(n, n)``` should be allowed to ```dask.array``` objects if the memory is overflowed or if the original array input by the user upon call to ```solve_ivp``` is a ```dask.array``` (or via a keyword option).

**Comment by rgommers** [comment]
In general, there will be quite a few scipy functions that just work when passing in a dask array rather than a numpy array. This isn't one of them, due to `np.asarray` calls and the use of sparse matrices (from a quick read, there may be more).

Keep in mind that while use of dask may help avoid a `MemoryError`, that doesn't mean the algorithm is going to be workably fast - often things need a few tweaks to perform well with dask.

This problem, where there is a spatial gradient connecting all points on a grid in principle, will not be a nice one for dask I'm afraid. This is more of a "buy/rent a bigger box" type problem.

> One potential solution might be to handle this error with a fallback to a dask.array if the package is available in the run-time environment.

I don't think we want that. Such logic is hard to maintain and debug. Things should not all of a sudden act differently is a bit more or less memory is available.

tl;dr this is not a good use case. I'd be in favor of making changes to code to make using dask.array inputs work instead of ndarray in places where that makes sense. Explicit use of `dask` within scipy code like you propose needs a much stronger case.

**Comment by rgommers** [comment]
I think this isn't really actionable at the moment, and the original question is answered. so I'll close this. thanks @guygma 

improvements that allow scipy functions to work better with dask without making scipy rely on dask are always welcome of course. removal of `asarray` in general is tricky though, because it may break backwards compat. perhaps soon we'll have an `asarray`-like duckarray function for that ....
